# Experimento numérico \# 5 - Tema: Solución de bloques para diferentes números de condición

**Responsable:** Dorely Morales Santiago

## 0. Objetivo

Este documento pretende mostar los resultados obtenidos con la implementación realizada, en el contexto del comportamiento del método de solución de bloques usando matrices de $10^2\times10^2$ o de $10^3\times10^3$ con número de condición de orden $10^0$ y $10^1$. 

**Nota:** El reporte general de resultados obtenidos con las implementaciones realizadas en este proyecto se puede consultar en el siguiente [vínculo](https://github.com/mno-2020-gh-classroom/ex-modulo-3-comp-matricial-svd-czammar/blob/master/results/Reporte_resultados.ipynb)

## 1. Consideraciones

Al respecto de los experimentos numéricos realizados para consolidar el presente reporte, tales se basan en las siguientes premisas:

**Las matrices con número de condición de orden $10^0$ y $10^1$ generan sistemas lineales bien condicionados (entre más cercano a 1 mejor) por lo que típicamente sus soluciones son rápidas de encontrar y tienen buena precisión.**

En este sentido, en particular para cada experimento realizado, se reportan según resulte aplicable:

* los parámetros empleados en la simulaciones, 
* las dimensiones de las matrices y vectores involucrados, así como el prodecimiento pseudo-aleatorio que les dio origen, 
* 1) el tiempo involucrado en correr los experimentos, 
* 2) número de condición de las matrices pseudo-aleatorias, y 
* 3) el error relativo obtenido, para la solución de un sistema lineal de la forma $Ax=b$, es decir el cociente

$$|| Ax - b ||_2 / || b ||_2$$
* 4) La norma de $A$ y de la matriz obtenida al rearmar A tras multiplicar las matrices resultantes de la aproximación de la descomposición SVD vía el algoritmo **One-Sided Jacobi**


### 1.1 Consideraciones sobre la infraestructura empleada

**Especificaciones de ambiente común de trabajo**

Para realizar el presente experimento numérico se ha empleado la imagen de docker basada en R del curso MNO 2020 (palmoreck/jupyterlab_r_kernel:1.1.0)

```
docker run --rm -v `pwd`:/datos --name jupyterlab_r_kernel_local -p 8888:8888 -d palmoreck/jupyterlab_r_kernel:1.1.0

```

*Nota:* el comando "-v \`pwd\`:/datos", permite montar el directorio actual en donde el usuario se encuentre situada en terminal como un volumen de la imagen de docker, dentro del directorio "/datos".

## 2. Experimento numérico

**Objetivo**

Modificar el número de condición de la matriz y las dimensiones de las matrices para notar el efecto que tiene sobre el error relativo y el tiempo de ejecución.

A tal respecto, se destaca que se realizó este experimento buscando probar cómo afecta el tiempo de ejecución el número de condición seleccionado.

**Cargamos código desarrollado previamente**

* **utils.R:** contiene las funciones auxiliares desarrolladas para el proyecto
* **00-load.R:** la implementación del método de eliminación por bloques, empleando la aproximación de la descomposición SVD vía el algoritmo **One-Sided Jacobi**.

In [1]:
## Instalamos paquetes
rm(list = ls())

paquetes <- c('matrixcalc','pracma')

instalar <- function(paquete) {
  if (!require(paquete,character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)) {
    install.packages(as.character(paquete), dependecies = TRUE, repos = "http://cran.us.r-project.org")
    library(paquete, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)
  }
}


lapply(paquetes, instalar)

[[1]]
NULL

[[2]]
NULL

In [2]:
## Cargamos paquetes necesarios
library("matrixcalc")
library("pracma")

#source("metadata.R")
source("utils.R")
source("00-load.R")

### 2.1 Experimento 5.1

Este experimento se basa en los siguientes parámetros:

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     |   $10^2$x$10^2$        |             |
| 2 |        Dimensión $b$       |      $10^2$            |             |
| 3 |         Tolerancia         |     $10^{-8}$          |             |
| 4 |          Maxsweep          |           5            |             |
| 5 |   Metodo pseudo-aleatorio $b$ |                        |        Obtenido con función rnorm    |
| 6 |   Metodo pseudo-aleatorio $A$ |                        |        Obtenido con función randortho    |
| 7 | Número de condición de $A$ |            $10^1$          |     Obtenido con función cond        |


**Código:** breve descrición de que hace el código y resultados.

In [36]:
set.seed(123)
n= 10**2
TOL = 10**-8

A <- diag(n)
A[1,1] <- 10
Q<-matrix(randortho(n,type = c("orthonormal")), ncol=n)
A<-Q%*%A%*%t(Q)
b = matrix(rnorm(n), ncol=1)

inicio<-Sys.time()
z<-eliminacion_bloques(A,b,2,10^-8,5)
print("#####################")
print("Error relativo")
norm(A%*%z-b,"2")/norm(b,"2")
print("#####################")
print("Número de condición")
cond(A)
print("#####################")
fin<-Sys.time()
fin-inicio

[1] "#####################"
[1] "Error relativo"


[1] 0.1120358

[1] "#####################"
[1] "Número de condición"


[1] 10

[1] "#####################"


Time difference of 1.588745 secs

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla:

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |           1.58 segs             |                                   |
| 2 |      Error relativo $||Ax-b||/||b||$     |           0.11203583997784           |                                   |


### 2.2  Experimento 5.2

Este experimento se basa en los siguientes parámetros:

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     |   $10^2$x$10^2$        |             |
| 2 |        Dimensión $b$       |      $10^2$            |             |
| 3 |         Tolerancia         |     $10^{-8}$          |             |
| 4 |          Maxsweep          |           5            |             |
| 5 |   Metodo pseudo-aleatorio $b$ |                        |        Obtenido con función rnorm    |
| 6 |   Metodo pseudo-aleatorio $A$ |                        |        Obtenido con función randortho    |
| 7 | Número de condición de $A$ |            $10^0$          |     Obtenido con función cond        |


**Código:** breve descrición de que hace el código y resultados.

In [37]:
set.seed(123)
n= 10**2
TOL = 10**-8

A <- diag(n)
Q<-matrix(randortho(n,type = c("orthonormal")), ncol=n)
A<-Q%*%A%*%t(Q)
b = matrix(rnorm(n), ncol=1)

inicio<-Sys.time()
z<-eliminacion_bloques(A,b,2,10^-8,5)
print("#####################")
print("Error relativo")
norm(A%*%z-b,"2")/norm(b,"2")
print("#####################")
print("Número de condición")
cond(A)
print("#####################")
fin<-Sys.time()
fin-inicio

[1] "#####################"
[1] "Error relativo"


[1] 2.210952e-15

[1] "#####################"
[1] "Número de condición"


[1] 1

[1] "#####################"


Time difference of 0.9170344 secs

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla:

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |           0.91 segs             |                                   |
| 2 |      Error relativo $||Ax-b||/||b||$     |             2.21095218644761e-15           |                                   |


### 2.1 Experimento 5.3

Este experimento se basa en los siguientes parámetros:

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     |   $10^3$x$10^3$        |             |
| 2 |        Dimensión $b$       |      $10^3$            |             |
| 3 |         Tolerancia         |     $10^{-8}$          |             |
| 4 |          Maxsweep          |           5            |             |
| 5 |   Metodo pseudo-aleatorio $b$ |                        |        Obtenido con función rnorm    |
| 6 |   Metodo pseudo-aleatorio $A$ |                        |        Obtenido con función randortho    |
| 7 | Número de condición de $A$ |            $10^1$          |     Obtenido con función cond        |


**Código:** breve descrición de que hace el código y resultados.

In [39]:
set.seed(123)
n= 10**3
TOL = 10**-8

A <- diag(n)
A[1,1] <- 10
Q<-matrix(randortho(n,type = c("orthonormal")), ncol=n)
A<-Q%*%A%*%t(Q)
b = matrix(rnorm(n), ncol=1)

inicio<-Sys.time()
z<-eliminacion_bloques(A,b,2,10^-8,5)
print("#####################")
print("Error relativo")
norm(A%*%z-b,"2")/norm(b,"2")
print("#####################")
print("Número de condición")
cond(A)
print("#####################")
fin<-Sys.time()
fin-inicio

[1] "#####################"
[1] "Error relativo"


[1] 0.03968741

[1] "#####################"
[1] "Número de condición"


[1] 10

[1] "#####################"


Time difference of 3.503863 mins

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla:

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |           3.50 mins             |                                   |
| 2 |      Error relativo $||Ax-b||/||b||$     |             0.0396874119847479           |                                   |


### 2.2  Experimento 5.4

Este experimento se basa en los siguientes parámetros:

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     |   $10^3$x$10^3$        |             |
| 2 |        Dimensión $b$       |      $10^3$            |             |
| 3 |         Tolerancia         |     $10^{-8}$          |             |
| 4 |          Maxsweep          |           5            |             |
| 5 |   Metodo pseudo-aleatorio $b$ |                        |        Obtenido con función rnorm    |
| 6 |   Metodo pseudo-aleatorio $A$ |                        |        Obtenido con función randortho    |
| 7 | Número de condición de $A$ |            $10^0$          |     Obtenido con función cond        |


**Código:** breve descripción de que hace el código y resultados.

In [38]:
set.seed(123)
n= 10**3
TOL = 10**-8

A <- diag(n)
Q<-matrix(randortho(n,type = c("orthonormal")), ncol=n)
A<-Q%*%A%*%t(Q)
b = matrix(rnorm(n), ncol=1)

inicio<-Sys.time()
z<-eliminacion_bloques(A,b,2,10^-8,5)
print("#####################")
print("Error relativo")
norm(A%*%z-b,"2")/norm(b,"2")
print("#####################")
print("Número de condición")
cond(A)
print("#####################")
fin<-Sys.time()
fin-inicio

[1] "#####################"
[1] "Error relativo"


[1] 7.463642e-15

[1] "#####################"
[1] "Número de condición"


[1] 1

[1] "#####################"


Time difference of 1.917876 mins

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla:

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |           1.91 mins             |                                   |
| 2 |      Error relativo $||Ax-b||/||b||$     |             7.46364171625283e-15           |                                   |


**Resumen de resultados obtenidos en experimentos**

Los resultados obtenidos, se compararán a través de la siguiente tabla:

| # | Experimento | Dimensiones de $A$ | Número de condición de $A$ | Tiempo de desempeño | Error relativo | Resultado |
|:-:|:-----------:|:-------------------:|----------------|----------------|-------|-------|
| 1 |     5.1     | $10^2$x$10^2$       |   $10^1$       | 1.58 segs      | 0.11203583997784      | rápido con error relativo alto      |
| 2 |     5.2     | $10^2$x$10^2$       |   $10^0$       | 0.91 segs      | 2.21095218644761e-15  | rápido y preciso      |
| 3 |     5.3     | $10^3$x$10^3$       |   $10^1$       | 3.50 mins      | 0.0396874119847479    | muy lento con error relativo alto     |
| 4 |     5.4     | $10^3$x$10^3$       |   $10^0$       | 1.91 mins      | 7.46364171625283e-15  | lento y preciso      |



## 3. Principales hallazgos

Al respecto, se destacan los siguientes hallazgos del experimento:

* Hallazgo 1: El tiempo de ejecución es mayor en matrices con número de condición de orden $10^1$ que de $10^0$. 
* Hallazgo 2: El error relativo en matrices de condición de orden de $10^0$ es prácticamente cero.
* Hallazgo 3: El error relativo en matrices de dimensión $10^2$x$10^2$ es significativamente superior al de matrices de $10^3$x$10^3$
